### ISSUE TITLE
refine the handling of datatypes

### ISSUE THREAD

**Issue by LinguList** [lingpy/lingpy#380]
Currently, we have a mess in our datatypes:

* alignments are `str`
* structure is `str`
* tokens are `list`

etc.

Especially for alignments and structure, the new code pieces (multitiers, lingrex, etc.) always convert between string and list. The problem is, that the alias system is allowing for too much freedom, and that the methods don't check for datatypes.

So we need to check how datatypes are handled:

1. `Alignments` class
2. `Partial`
3. `LexStat`

and in the new future addons

4. `CoPaR`
5. `MultiTiers`

I suggest the following

name | type | note
--- | --- 
alignment | list | `alignments` keyword needs to be checked on default, whether it is a string or a list and change to list
structure | list | same as for `alignments`, we now assume tab-separated string
morphemes | list

There should be some documentation that

1. presents the datatypes
2. explains where they play a role, and how we access them
3. explains that they are automatically converted, if users choose another column name

Since we now add the keywords to the calls, like

```python
>>> alm = Alignments(segments="tokens", ref="cogids", fuzzy=True)
```

we can afford that little extra time to check that the data is rendered internally in a consistent way.

**Comment by SimonGreenhill** [comment]
As a first step, what about using python's new [type hints](https://www.python.org/dev/peps/pep-0484/) to annotate types? this can then be checked on travis with [mypy](http://mypy-lang.org/)

**Comment by LinguList** [comment]
Thanks, will try and look into that!

**Comment by xrotwang** [comment]
Type hints may be a good idea, although I'd prefer buying into this only when abandoning py2 - otherwise we have to deal with stuff like this https://stackoverflow.com/a/35230792

**Comment by LinguList** [comment]
I am experimenting on simple tests that I may prepare for PR:

* load an alignment object, allow for keywords for all those aspects where it is crucial what type the data has
* check if the datatype is correctly interpreted
* try to change it if it is not
* raise an error otherwise

I think I can easily pin down what data would need to be represented in what form. The problem is the alias system, and that it may occur as a use case that we want to use different column names for it. I feel that this needs to be supported by lingpy, to allow for flexibility and future development, but I also feel that we should make this behavior less implicit (using the rc file), but more explicit, by maybe creating type checks and corrections for the four major regular datatypes (lists of strings, lists of integers, integers, and strings), but the users will have to be explicit about the type when loading their wordlist.

**Comment by LinguList** [comment]
example is [here](https://github.com/LinguList/lingpy/commit/5e8adf5baf89481f518a1262ffa78f5fa735ae03#diff-e3ffb5de94e01fedeed0a88463b107cfL613), but I'm working to enhance this. If there's some idea of how to modularize this more, or a general recommendation (like "put this already in wordlist"), this would be helpful, but I think we can afford three more lines of code in the Alignments package for the time being.

**Comment by SimonGreenhill** [comment]
py2's EOL in 2020, time to move on!

**Comment by LinguList** [comment]
I just started with a basic structure of our lingpy functions, e.g., a base list that is space-segmented strings, initiated by splitting on space:

```python
class blist(list):

    def __init__(self, iterable):

        list.__init__(self, iterable.split())

    def __str__(self):

        return ' '.join(self)
```

for the integer list (also needed), this would simply be:

```python
class ilist(list):

    def __init__(self, iterable):

        list.__init__(self, [int(x) for x in iterable.split()])

    def __str__(self):

        return ' '.join([str(x) for x in self])
```

**Comment by LinguList** [comment]
Yes, good point, I just also realized that this is needed.

And I realized, I'll have to add few more datatypes:

* blist : list of strings
* ilist : list of integers
* flist : list of floats
* mlist : list that has a nested representation (if we have "+" for morphemes plus spacesegmentation) as well as a normal represetnation. If we allow for two ways to access this, e.g., `mlist.n` vs. `mlist` proper, we can again save a lot of code lingpy-internal...

**Comment by xrotwang** [comment]
Also, this is already py3 only. For py2 support, you might want to inherit from `clldutils.misc.UnicodeMixin` and overwrite `__unicode__`. But maybe it's time to drop py2?

**Comment by LinguList** [comment]
I am not against keeping py2 now, as it is straightforward to go for this, right?

Although I just saw that the generator_stop introduced in Python 3.7 is no longer easy to support in 2.7 and it breaks lingpy when used by py3.7.

**Comment by LinguList** [comment]
see here:

* https://github.com/lingpy/lingpy/commit/f643c95a008edfc64acc09ea3cbe53ae9acec5a5

**Comment by xrotwang** [comment]
Also, for lists with elements of fixed numeric type, we may want to inherit from [`array`](https://pymotw.com/2/array/) instead.

**Comment by xrotwang** [comment]
Yes, I guess the more py3 versions we want to support, the more py2 support is going to be an excess burden. So maybe drop it already?

**Comment by LinguList** [comment]
Okay. I'd say: if we go and drop, we should go for version 2.7 with that very version of lingpy, to make a clearer cut. 

It'll cost some time to include these things. But it will save a lot of code, already now, I have huge problems in writing "tokens2morphemse(tokens)" all the time, just because I need a doubly nested list. 

Should we start by just writing the lingpy.basictypes or lingpy.types script and discuss the different datatypes there, along with tests and best practices?

**Comment by xrotwang** [comment]
Yes, let's do some experimenting, and I'd vote for `basictypes` since otherwise shadowing a standard lib module may be happen too easily.

Btw.: if typing long function names creates "huge problems", use PyCharm :)

**Comment by LinguList** [comment]
:) yes, but typing a lot is not the problem, it's the visibility of code with all those nestings...
